import gensim
import gensim.corpora as corpora
from gensim.corpora import Dictionary
from gensim.models.coherencemodel import CoherenceModel
from gensim.models.ldamodel import LdaModel

from pprint import pprint

import spacy

import pickle
import re 
import pyLDAvis
import pyLDAvis.gensim

import matplotlib.pyplot as plt 
import pandas as pd
___

import numpy
numpy.random.bit_generator = numpy.random._bit_generator

### fully uninstalled, reinstalled numpy, restarted kernel, finally 
import gensim
from gensim import corpora
#### about 18 cells down runs, creates:
###### Dictionary(54130 unique tokens: ['0', '0frac', '0mathbf', '0ttfrac', '1']...)

import numpy
numpy.random.bit_generator = numpy.random._bit_generator

import numpy as np
rng = np.random.default_rng(12345)
print(rng)
Generator(PCG64)
rfloat = rng.random()
rfloat
0.22733602246716966
type(rfloat)

random.Generator.bit_generator

with help from

https://shichaoji.com/tag/topic-modeling-python-lda-visualization-gensim-pyldavis-nltk/

In [1]:
from helper import *

ModuleNotFoundError: No module named 'helper'

In [2]:
import warnings

In [3]:
warnings.filterwarnings('ignore')

In [4]:
import nltk
nltk.download('stopwords')
nltk.download('wordnet')

C:\Users\Owner\Anaconda3\lib\site-packages\sklearn\feature_extraction\image.py:167: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  dtype=np.int):
C:\Users\Owner\Anaconda3\lib\site-packages\sklearn\linear_model\least_angle.py:30: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.o

True

In [5]:
from nltk.corpus import stopwords 
from nltk.stem.wordnet import WordNetLemmatizer
import string

stopwords = set(stopwords.words('english'))
punctuation = set(string.punctuation) 
lemmatize = WordNetLemmatizer()

def cleaning(article):
    one = " ".join([i for i in article.lower().split() if i not in stopwords])
    two = "".join(i for i in one if i not in punctuation)
    three = " ".join(lemmatize.lemmatize(i) for i in two.split())
    return three

In [6]:
import pandas as pd
import numpy as np
import re
import json
import glob

In [7]:
folder = (glob.glob('../data/english_topic_full_content/*'))

In [8]:
article_list = []
for file in folder:
    with open(file, 'r') as f:
        article = json.load(f)
    article_list.append(article)

In [9]:
articles = pd.DataFrame(article_list).replace(r'^\s*$', np.nan, regex = True)
articles.info()

#ignore error - dataframe still runs

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 319 entries, 0 to 318
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   topic          319 non-null    object
 1   english_title  243 non-null    object
 2   english        243 non-null    object
dtypes: object(3)
memory usage: 7.6+ KB


In [10]:
articles = articles.dropna()

In [11]:
articles.head(200)

topic english_title  \
0    Acceleration  Acceleration   
1      acceptance    Acceptance   
4       Aerospace     Aerospace   
5    Agricultural   Agriculture   
6     Agriculture   Agriculture   
..            ...           ...   
262      Shopping      Shopping   
264        Social        Social   
265       Society       Society   
266     Sociology     Sociology   
267      software      Software   

                                               english  
0    In mechanics, acceleration is the rate of chan...  
1    Acceptance in human psychology is a person's a...  
4    Aerospace is a term used to collectively refer...  
5    Agriculture is the science, art and practice o...  
6    Agriculture is the science, art and practice o...  
..                                                 ...  
262  Shopping is an activity in which a customer br...  
264  Social organisms, including humans, live colle...  
265  A society is a group of individuals involved i...  
266  Sociology is the study of society, human socia...  
267  Software is a collection of instructions and d...  

[200 rows x 3 columns]

In [12]:
text = articles.applymap(cleaning)['english']
text_list = [i.split() for i in text]
len(text_list)

243

In [13]:
text_list[0]

['mechanic',
 'acceleration',
 'rate',
 'change',
 'velocity',
 'object',
 'respect',
 'time',
 'acceleration',
 'vector',
 'quantity',
 'in',
 'magnitude',
 'direction',
 'orientation',
 'object',
 'acceleration',
 'given',
 'orientation',
 'net',
 'force',
 'acting',
 'object',
 'magnitude',
 'object',
 'acceleration',
 'described',
 'newton',
 'second',
 'law',
 'combined',
 'effect',
 'two',
 'cause',
 'net',
 'balance',
 'external',
 'force',
 'acting',
 'onto',
 'object',
 '—',
 'magnitude',
 'directly',
 'proportional',
 'net',
 'resulting',
 'force',
 'object',
 'mass',
 'depending',
 'material',
 'made',
 '—',
 'magnitude',
 'inversely',
 'proportional',
 'object',
 'massthe',
 'si',
 'unit',
 'acceleration',
 'metre',
 'per',
 'second',
 'squared',
 'm⋅s−2',
 '2',
 'displaystyle',
 'tfrac',
 'operatorname',
 'm',
 'operatorname',
 's',
 '2',
 'example',
 'vehicle',
 'start',
 'standstill',
 'zero',
 'velocity',
 'inertial',
 'frame',
 'reference',
 'travel',
 'straight',
 'li

#progress log unnecessary

from time import time
import logging

logging.basicConfig(format='%(asctime)s : %(levelname)s : %(message)s', level=logging.INFO,
                   filename='running.log',filemode='w')

In [15]:
# Importing Gensim
import gensim
from gensim import corpora

# Creating the term dictionary of our courpus, where every unique term is assigned an index. dictionary = corpora.Dictionary(doc_clean)
dictionary = corpora.Dictionary(text_list)
dictionary.save('dictionary.dict')
print(dictionary)

Dictionary(54130 unique tokens: ['0', '0frac', '0mathbf', '0ttfrac', '1']...)


In [16]:
# Converting list of documents (corpus) into Document Term Matrix using dictionary prepared above.
doc_term_matrix = [dictionary.doc2bow(doc) for doc in text_list]
corpora.MmCorpus.serialize('corpus.mm', doc_term_matrix)

len(doc_term_matrix)
print(doc_term_matrix[100])

[(8, 1), (21, 1), (23, 1), (27, 3), (29, 10), (30, 2), (31, 1), (32, 2), (42, 5), (47, 4), (50, 2), (51, 3), (52, 2), (63, 3), (64, 2), (70, 1), (75, 2), (76, 1), (88, 1), (103, 2), (108, 1), (109, 2), (110, 1), (123, 1), (131, 2), (145, 1), (147, 6), (153, 1), (155, 2), (160, 1), (164, 5), (168, 1), (177, 1), (180, 1), (191, 1), (193, 1), (199, 2), (213, 1), (215, 5), (223, 1), (229, 3), (233, 1), (237, 8), (254, 2), (261, 3), (268, 3), (269, 1), (272, 9), (274, 6), (281, 1), (286, 4), (288, 3), (292, 2), (300, 2), (303, 2), (316, 1), (328, 3), (329, 1), (335, 1), (343, 1), (348, 1), (349, 1), (350, 3), (352, 2), (353, 2), (374, 8), (380, 1), (383, 17), (394, 3), (399, 3), (401, 1), (408, 4), (409, 1), (419, 3), (429, 2), (468, 2), (477, 1), (482, 3), (492, 1), (494, 6), (496, 1), (505, 2), (518, 2), (520, 1), (522, 1), (532, 4), (546, 2), (556, 2), (561, 1), (564, 2), (565, 1), (571, 1), (572, 1), (573, 1), (575, 1), (576, 2), (584, 1), (585, 2), (608, 3), (609, 2), (610, 1), (625, 1

In [19]:
start = time()
# Creating the object for LDA model using gensim library
Lda = gensim.models.ldamodel.LdaModel

# Running and Trainign LDA model on the document term matrix.
ldamodel = Lda(doc_term_matrix, num_topics=10, id2word = dictionary, passes=50)
print('used: {:.2f}s'.format(time()-start))

used: 262.84s


In [20]:
print(ldamodel.print_topics(num_topics=2, num_words=4))

[(5, '0.006*"world" + 0.005*"history" + 0.005*"military" + 0.005*"country"'), (8, '0.008*"study" + 0.007*"science" + 0.007*"research" + 0.007*"social"')]


In [21]:
for i in ldamodel.print_topics(): 
    for j in i: print(j)

0
0.006*"food" + 0.006*"art" + 0.005*"woman" + 0.005*"human" + 0.005*"rhetoric" + 0.005*"also" + 0.004*"massage" + 0.004*"feminist" + 0.004*"entertainment" + 0.004*"form"
1
0.008*"technology" + 0.006*"design" + 0.006*"innovation" + 0.006*"system" + 0.006*"used" + 0.006*"process" + 0.005*"engineering" + 0.005*"use" + 0.005*"computer" + 0.004*"also"
2
0.008*"education" + 0.005*"religion" + 0.004*"school" + 0.004*"child" + 0.004*"ancient" + 0.004*"century" + 0.004*"age" + 0.004*"area" + 0.004*"robot" + 0.004*"bc"
3
0.010*"greek" + 0.009*"china" + 0.008*"wikipedia" + 0.007*"greece" + 0.005*"island" + 0.005*"world" + 0.004*"myth" + 0.004*"country" + 0.004*"article" + 0.004*"chinese"
4
0.012*"theory" + 0.006*"one" + 0.005*"time" + 0.004*"human" + 0.004*"concept" + 0.004*"belief" + 0.004*"also" + 0.004*"philosophy" + 0.004*"study" + 0.003*"example"
5
0.006*"world" + 0.005*"history" + 0.005*"military" + 0.005*"country" + 0.005*"internet" + 0.004*"egypt" + 0.004*"also" + 0.004*"communication" +

In [22]:
ldamodel.save('topic.model')

### start here

In [23]:
from gensim.models import LdaModel
loading = LdaModel.load('topic.model')

In [24]:
print(loading.print_topics(num_topics=2, num_words=4))

[(3, '0.010*"greek" + 0.009*"china" + 0.008*"wikipedia" + 0.007*"greece"'), (8, '0.008*"study" + 0.007*"science" + 0.007*"research" + 0.007*"social"')]


In [25]:
def pre_new(doc):
    one = cleaning(doc).split()
    two = dictionary.doc2bow(one)
    return two

In [ ]:
pre_new('new article that to be classified by trained model!')

In [ ]:
belong = loading[(pre_new('new article that to be classified by trained model!'))]
belong

In [ ]:
new = pd.DataFrame(belong,columns=['id','prob']).sort_values('prob',ascending=False)
new['topic'] = new['id'].apply(loading.print_topic)
new

In [ ]:
new['topic']

In [ ]:
import gensim
from gensim import corpora

ran next cell again

In [31]:
import pyLDAvis.gensim
import gensim
pyLDAvis.enable_notebook()
import pyLDAvis.gensim_models as gensimvis

###### conda install -c conda-forge pyldavis finally did it

In [28]:
import pyLDAvis
#import pyLDAvis.gensim_models as gensimvis  COMMENTED OUT CAUSE DIDN'T NEED HERE
pyLDAvis.enable_notebook()

Successfully built sklearn
Installing collected packages: funcy, sklearn, pandas, pyLDAvis
  Found existing installation: pandas 0.25.1
    Uninstalling pandas-0.25.1:
      Successfully uninstalled pandas-0.25.1
ERROR: Could not install packages due to an EnvironmentError: [WinError 5] Access is denied: 'c:\\users\\owner\\anaconda3\\lib\\site-packages\\~andas\\io\\msgpack\\_packer.cp37-win_amd64.pyd'
Consider using the `--user` option or check the permissions.

In [32]:
vis_data = gensimvis.prepare(ldamodel, doc_term_matrix, dictionary)
pyLDAvis.display(vis_data)

C:\Users\Owner\Anaconda3\lib\site-packages\joblib\numpy_pickle.py:104: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.
  pickler.file_handle.write(chunk.tostring('C'))
C:\Users\Owner\Anaconda3\lib\site-packages\joblib\numpy_pickle.py:104: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.
  pickler.file_handle.write(chunk.tostring('C'))


##### worked anyway?  .. then it worked

In [17]:
d = gensim.corpora.Dictionary.load('dictionary.dict')
c = gensim.corpora.MmCorpus('corpus.mm')
lda = gensim.models.LdaModel.load('topic.model')

###### then, pip install pyLDAvis==2.1.2 or not

In [18]:
import gensim
from gensim import corpora

##### after got to end, came back and ran this again
The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
--> 232         use_numexpr = use_numexpr and _bool_arith_check(op_str, a, b)
    233         if use_numexpr:
    234             # error: "None" not callable

#Train LDA model.
from gensim.models import LdaModel

#Set training parameters.
num_topics = 10
chunksize = 2000
passes = 20
iterations = 400
eval_every = None  # Don't evaluate model perplexity, takes too much time.

#Make a index to word dictionary.
temp = dictionary[0]  # This is only to "load" the dictionary.
id2word = dictionary.id2token

model = LdaModel(
    corpus=corpus,
    id2word=id2word,
    chunksize=chunksize,
    alpha='auto',
    eta='auto',
    iterations=iterations,
    num_topics=num_topics,
    passes=passes,
    eval_every=eval_every
)

# HERE

In [ ]:
articles = pyLDAvis.gensim.prepare(lda, c, d)
articles

In [ ]:
articles = pyLDAvis.gensim.prepare(optimal_model, corpus, id2word)
articles

In [ ]:
pyLDAvis.enable_notebook()
articles = pyLDAvis.gensim.prepare(lda, c, d)
articles

In [ ]:
from pyLDAvis import gensim

###### worked anyway??

In [ ]:
pyLDAvis.save_html(articles,'vis.html')

In [ ]:
display(HTML(filename = 'vis.html'))

In [ ]:
import codecs
import webbrowser

In [ ]:
webbrowser.open('vis.html')

trying with pyLDAvis.enable_ntebook(local=True) worked for me - some forum